lwt/xen_api_lwt_unix.ml

```diff
@@ -87,24 +87,34 @@ module Lwt_unix_IO = struct
 
 		if ssl then begin
 			let fd = Lwt_unix.socket domain Unix.SOCK_STREAM 0 in
-			try_lwt
-				lwt () = Lwt_unix.connect fd sockaddr in
-				lwt sock = Lwt_ssl.ssl_connect fd sslctx in
-				let ic = Lwt_ssl.in_channel_of_descr sock in
-				let oc = Lwt_ssl.out_channel_of_descr sock in
-				return (Ok (((fun () -> Lwt_ssl.close sock), ic), ((fun () -> Lwt_ssl.close sock), oc)))
-			with e ->
-				return (Error e)
+			Lwt.catch
+				(fun () ->
+					 Lwt.catch (fun () ->
+							 Lwt_unix.connect fd sockaddr
+						 ) (fun e ->
+							 Lwt_unix.close fd >>= fun () -> Lwt.fail e
+						 )
+					 >>= fun () ->
+					 Lwt_ssl.ssl_connect fd sslctx >>= fun sock ->
```

Comment:
What if Lwt_ssl.ssl_connect fails?

Reply:
It would be caught by the outer catch?

Reply:
Yes, which would return an Error to the caller, but the fd is leaked. I think if we move the ssl_connect inside the inner catch then it would avoid that leak.

Reply:
I have pushed a fixup. We should fix this on the master branch, too.